site/src/pages/types/[entityTypeId].page.tsx

```diff
@@ -0,0 +1,50 @@
+import { GetServerSideProps, NextPage } from "next";
+import { connectToDatabase } from "../../lib/mongodb";
+import { EntityType } from "../../lib/model/entityType.model";
+import { FRONTEND_URL } from "../../lib/config";
+
+const EntityTypeIdPage: NextPage = () => <div />;
+
+export default EntityTypeIdPage;
+
+export const getServerSideProps: GetServerSideProps<
+  {},
+  {
+    entityTypeId: string;
+  }
+> = async (context) => {
+  const { req, res } = context;
+
+  const { db } = await connectToDatabase();
+
+  const entityTypeId = context.query.entityTypeId as string;
+  const entityType = await EntityType.getById(db, { entityTypeId });
+
+  const jsonFlag = new URL(`${FRONTEND_URL}${req.url}`).searchParams.get(
+    "json",
+  );
+
+  if (jsonFlag != null || req.headers.accept?.includes("application/json")) {
+    if (entityType) {
+      res.setHeader("Content-Type", "application/json; charset=utf-8");
+      res.write(JSON.stringify(entityType.schema, undefined, 2));
+    } else {
+      res.statusCode = 404;
+      res.write("Schema not found.");
+    }
+    res.end();
+  } else if (entityType) {
+    return {
+      redirect: {
+        permanent: false,
+        destination: `/${entityType.schema.author}/types/${entityType.schema.title}`,
+      },
+    };
+  } else {
+    return {
+      notFound: true,
+    };
```

Comment:
We return a short text message in one case and a 404 page in another. It’d be great to align them.

If we want `/types/xyz` to act as an API, we can move it into the `/api/` folder and add a [rewrite](https://nextjs.org/docs/api-reference/next.config.js/rewrites). This will give `res.json()` as part of `NextApiHandler`, so we won’t have to worry about headers.

Reply:
I think for things that aren't expected to return web pages, we don't need to return a 404 page.

I'd like to keep the URL for types as `/types/xxx` rather than adding an intervening `/api/`. Keeping it here makes it clear what's going on, rather than having to achieve that through redirects elsewhere.

I don't think being able to remove a line setting the header is worth the additional complexity.

Reply:
IC.

Next renders `return { notFound: true }` as an HTML page. So we probably want to replace that with `res.statusCode = 404; res.write("Schema not found."); res.end();` for consistency with another `if/else` branch. That does not have to be done now though.

Reply:
> Next renders return { notFound: true } as an HTML page

Right, but the user only reaches that if they're in the HTML branch - the intention of this route is that you either
1) get JSON (and a plain 404) if you've requested JSON, otherwise
2) sends you to the HTML page for the type, or a 404 page if it doesn't exist

I'll get this in and we can refine afterwards - thank you!

Reply:
Ah right – I forgot about `jsonFlag` we have in the first `if`. All makes sense not, sorry for the noise! 🤦